third_party/gtp_ovs/ovs-gtp-patches/2.15/0024-QFI-support-in-OVS.patch

```diff
@@ -0,0 +1,878 @@
+From: Prabina Pattnaik <prabinak@wavelabs.ai>
+Date: Sat, 30 Apr 2022 11:42:03 +0000
+Subject: [PATCH 24/24] QFI support in OVS
+
+Signed-off-by: Prabina Pattnaik <prabinak@wavelabs.ai>
+---
+ datapath/linux/compat/gtp.c               | 145 ++++++++++++----------
+ datapath/linux/compat/include/linux/gtp.h |   1 +
+ include/openvswitch/match.h               |   2 +
+ include/openvswitch/meta-flow.h           |  17 +++
+ include/openvswitch/ofp-actions.h         |  13 ++
+ include/openvswitch/packets.h             |   3 +-
+ lib/flow.c                                |   6 +
+ lib/match.c                               |  17 +++
+ lib/meta-flow.c                           |  19 +++
+ lib/meta-flow.xml                         |   9 ++
+ lib/nx-match.c                            |   2 +-
+ lib/odp-util.c                            |  13 +-
+ lib/ofp-actions.c                         |  73 +++++++++++
+ lib/packets.h                             |  11 +-
+ tests/ofproto.at                          |   4 +-
+ tests/system-layer3-tunnels.at            | 108 ++++++++++++++++
+ tests/tunnel.at                           |  21 ++++
+ 17 files changed, 394 insertions(+), 70 deletions(-)
+
+diff --git a/datapath/linux/compat/gtp.c b/datapath/linux/compat/gtp.c
+index a7033031d..f2e0ca4bf 100644
+--- a/datapath/linux/compat/gtp.c
++++ b/datapath/linux/compat/gtp.c
+@@ -113,18 +113,39 @@ static int gtp_rx(struct sock *sk, struct gtp_dev *gtp, struct sk_buff *skb,
+ 			udp_tun_rx_dst(skb, sk->sk_family, TUNNEL_KEY, tid, opts_len);
+ #endif
+ 		netdev_dbg(gtp->dev, "attaching metadata_dst to skb, gtp ver %d hdrlen %d\n", gtp_version, hdrlen);
+-		if (unlikely(opts_len)) {
+-			struct gtpu_metadata *opts = ip_tunnel_info_opts(&tun_dst->u.tun_info);
+-			struct gtp1_header *gtp1 = (struct gtp1_header *)(skb->data + sizeof(struct udphdr));
+-
+-			opts->ver = GTP_METADATA_V1;
+-			opts->flags = gtp1->flags;
+-			opts->type = gtp1->type;
+-			netdev_dbg(gtp->dev, "recved control pkt: flag %x type: %d\n", opts->flags, opts->type);
+-			tun_dst->u.tun_info.key.tun_flags |= TUNNEL_GTPU_OPT;
+-			tun_dst->u.tun_info.options_len = opts_len;
+-			skb->protocol = 0xffff;         // Unknown
++                if (unlikely(opts_len)) {
++                    struct gtpu_metadata *opts = ip_tunnel_info_opts(&tun_dst->u.tun_info);
++                    struct gtp1_header *gtp1 = (struct gtp1_header *)(skb->data + sizeof(struct udphdr));
++
++                    opts->ver = GTP_METADATA_V1;
++                    opts->flags = gtp1->flags;
++                    opts->type = gtp1->type;
++                    netdev_dbg(gtp->dev, "recved control pkt: flag %x type: %d\n", opts->flags, opts->type);
++                    tun_dst->u.tun_info.key.tun_flags |= TUNNEL_GTPU_OPT;
++                    tun_dst->u.tun_info.options_len = opts_len;
++                    skb->protocol = 0xffff;         // Unknown
++                }
++
++		if (likely(type == GTP_TPDU)){
```

Comment:
you need to set `opts_len` to sizeof (struct gtpu_metadata), before you can use optional data for QFI case.

Reply:
Ok, I have set the ``` opts_len``` to sizeof ``` struct gtpu_metadata ``` if qfi value is present.